Ddci phase1a

jasmin-j opened the pull request:
After lots of hours debugging. I fixed the root cause of the problem.
The CA descriptor PIDs (ECMs) weren't subscribed by the dvbcam.
I added the ECM pid add/remove handling to dvbcam. There is still some code missing, but as a first step it is ok.
I am not sure if there would have been a simpler solution by using the already existing table handlers, but the current solution subscribes only those ECMs which are required by the CAM to descramble the CAIDs the CAM can support.

NOTE: The stream can't be displayed by VLC or other programs, because there are still some PIDs missing. But the stream can be displayed with ffplay, so the CAM is now fully decrypting the stream.

What is missing to get this fully working are the following PIDs in the output stream:
0x00	PAT
0x11	SDT/BAT
0x???	PMT
This PIDs are sent to the CAM and also are sent back, but they are not streamed to the client. I have no clue where and how I can enable this.
@perexg, please tell me where and how I can subscribe these PIDs, so that they are send to the client.

Trujulu commented:
:o awesome

jasmin-j commented:
> but the current solution subscribes only those ECMs which are required by the CAM to descramble the
> CAIDs the CAM can support.

Maybe it would be better to send only the ECM PIDs of the currently selected service instead of all the PIDs the CAM supports.
@perexg, how can I determine in the dvbcam_service_start function, which PIDs are required for the actual service only?

perexg commented:
@jasmin-j : You already filter only ECM PIDs for given service in your code (es_type == SCT_CA and .use flag). You probably mean EMM PIDs (from CAT table). I am working on it right now.

perexg commented:
I pushed your changes with my 'fixes' on top. Basically, we have already PID management routines in tvh (mpegts_apid_t). I added the limit back (per cam now) - we can discuss if it's the best solution.

I reimplemented CAT (EMM) PID management and fixed some bugs (and maybe added others - sorry - no hw to test here). Note that I have to remove TSS_NO_DESCRAMBLER settings - it does not work at all, if you have other descrambling clients enabled. If you have only one DVBCAM/DDCI client and it is not allocated, dr->dr_ca_count should be zero, thus descrambler_descramble() will return 0 and ts_recv_packet1() should set this flag (TSS_NO_DESCRAMBLER) in this case.

herrnst commented:
@perexg @jasmin-j Latest HEAD makes things way worse than before - my CAM (at adapter7/{ca0|sec0}) now even isn't considered anymore when switching to a scrambled channel. On startup or activation via the WebUI, the CAM is initialised, identified as AlphaCrypt and the ddci code initialises the device "as usual", but nothing happens furthermore. Log output got reduced to this (disabled the noisy mpegts debug):

    2017-11-20 19:44:42.799 [   INFO]:mpegts: 354MHz in Unitymedia NRW - tuning on ST STV0367 DDB DVB-C/T : DVB-C #0
    2017-11-20 19:44:42.799 [  TRACE]:dvbcam: start service 0x55590149a930
    2017-11-20 19:44:42.799 [   INFO]:subscription: 000A: "192.168.0.120 [  | Kodi Media Center ]" subscribing on channel "Fox HD", weight: 150, adapter: "ST STV0367 DDB DVB-C/T : DVB-C #0", network: "Unitymedia NRW", mux: "354MHz", provider: "Unitymedia", service: "Fox HD", profile="htsp", hostname="192.168.0.120", username="192.168.0.120", client="Kodi Media Center"
    2017-11-20 19:44:43.501 [  TRACE]:descrambler: CAT data (len 42)
    2017-11-20 19:44:43.501 [  TRACE]:descrambler: [hex dump]
    2017-11-20 19:44:43.700 [  TRACE]:dvbcam: cannot find active service entry

and then nothing more. No deadlock or so though since I can unsubscribe the service, use the WebUI normally etc., thats just everything that's printed.

jasmin-j commented:
> Latest HEAD makes things way worse than before

I will have a look tomorrow.

jasmin-j commented:
> we have already PID management routines in tvh (mpegts_apid_t).

mpegts_apids_t -> This is a really cool class!

> I added the limit back (per cam now) - we can discuss if it's the best solution.

I will think about it. THX for implementing!

> I reimplemented CAT (EMM) PID management and fixed some bugs (and maybe added others - sorry - no
> hw to test here).

Maybe you should make a test or devel branch, so that those bigger changes doesn't break existing functionality.

> Note that I have to remove TSS_NO_DESCRAMBLER settings

I implemented it according to an eMail advice from you, but your new solution should work. I will test that.

@perexg: Code review of your changes:
```
mpegts_input.c:
  for (es = TAILQ_FIRST(&s->s_components); es != NULL; es = next) {
    next = TAILQ_NEXT(es, es_link);
    if (es->es_type != SCT_CAT) continue;
    for (c = LIST_FIRST(&es->es_caids); c != NULL; c = cn) {
      cn = LIST_NEXT(c, link);
      if (cn->delete_me) {      <--- shouldn't this be 'c' instead of cn?
        LIST_REMOVE(c, link);
        free(c);
      }
      if (es->es_delete_me)               <--- This 'if' should be outside the loop
        service_stream_destroy(s, es);         otherwise the next iteration will
    }                                          access the deleted 'es'!
  }

dvbcam.c:
  TAILQ_FOREACH(as, &dvbcam_active_services, global_link) {
    if (!as->is_ddci) continue;
    if (as->caid == 0) continue;
    if (((mpegts_service_t *)as->td_service)->s_dvb_mux != mux) continue;
    if (len == as->cat_data_len) continue;               <--- Are you sure this is clever?
    if (memcmp(data, as->cat_data, len) == 0) continue;       If the new data has the same length,
                                                              but a different content the new data
                                                              is ignored
```

And it crashed:
```
2017-11-21 00:45:56.836 [ NOTICE] ddci: CAM ci0 assigned to 0x7fd043566190
2017-11-21 00:45:56.836 [   INFO] subscription: 0001: "HTTP" subscribing on channel "ORF1 HD", weight: 100, adapter: "STV090x Multistandard #1 : DVB-S #0", network: "DVB-S Network", mux: "11302.75H", provider: "ORF", service: "ORF1 HD", profile="pass", hostname="192.168.23.40", client="Videos/1.2.2"
2017-11-21 00:45:57.463 [  ERROR] mpegts: 11302.75H in DVB-S Network - open PID 0001 (1) failed, dupe sub (owner 0x7fd043566190)
2017-11-21 00:45:57.531 [  ERROR] mpegts: 11302.75H in DVB-S Network - open PID 0001 (1) failed, dupe sub (owner 0x7fd043566190)
2017-11-21 00:45:58.211 [  ALERT] CRASH: Signal: 11 in PRG: ./build.linux/tvheadend (4.3-670~geaa7bf5) [8dd3015340e92a3d42f46b4208b23c9c673403df] CWD: /mnt/nfs/hoppel_home_hdd/jasmin/tvheadend/tvheadend  
2017-11-21 00:45:58.211 [  ALERT] CRASH: Fault address (nil) (Address not mapped)
2017-11-21 00:45:58.211 [  ALERT] CRASH: Loaded libraries: /usr/lib/libdvben50221.so /usr/lib/libdvbapi.so /usr/lib/libucsi.so /lib/x86_64-linux-gnu/libssl.so.1.0.0 /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 /lib/x86_64-linux-gnu/libz.so.1 /lib/x86_64-linux-gnu/libpcre.so.3 /usr/lib/liburiparser.so.1 /usr/lib/x86_64-linux-gnu/libavahi-common.so.3 /usr/lib/x86_64-linux-gnu/libavahi-client.so.3 /lib/x86_64-linux-gnu/libdbus-1.so.3 /lib/x86_64-linux-gnu/libdl.so.2 /lib/x86_64-linux-gnu/libpthread.so.0 /lib/x86_64-linux-gnu/libm.so.6 /lib/x86_64-linux-gnu/librt.so.1 /lib/x86_64-linux-gnu/libc.so.6 /lib64/ld-linux-x86-64.so.2  
2017-11-21 00:45:58.211 [  ALERT] CRASH: Register dump [23]: 0000000000000000202020202020202000007fd02c05244000007fd02c05246c00007fd04238903000007fd02c00091000007fd01cde99c000007fd01cde970000007fd02c052420ffff802fd3fadbe000007fd01cde86900000000000000010000000000000004c00007fd02c052420000000000000000000007fd01cde7ef800007fd03db4c596000000000001020200000000000000330000000000000004000000000000000efffffffe7ffbba110000000000000000
2017-11-21 00:45:58.211 [  ALERT] CRASH: STACKTRACE
2017-11-21 00:45:58.322 [  ALERT] CRASH: /tvheadend/tvheadend/src/trap.c:148 0x7fd0404cda95 0x7fd040319000
2017-11-21 00:45:58.355 [  ALERT] CRASH: ??:0 0x7fd03e3a7330 0x7fd03e397000
2017-11-21 00:45:58.389 [  ALERT] CRASH: ??:0 0x7fd03db4c596 0x7fd03dac0000
2017-11-21 00:45:58.422 [  ALERT] CRASH: /tvheadend/tvheadend/src/descrambler/dvbcam.c:555 0x7fd040614f79 0x7fd040319000
2017-11-21 00:45:58.456 [  ALERT] CRASH: /tvheadend/tvheadend/src/descrambler/caclient.c:346 0x7fd040573dd8 0x7fd040319000
2017-11-21 00:45:58.489 [  ALERT] CRASH: /tvheadend/tvheadend/src/descrambler/descrambler.c:1484 0x7fd040572829 0x7fd040319000
2017-11-21 00:45:58.523 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/dvb_psi.c:1050 0x7fd040597240 0x7fd040319000
2017-11-21 00:45:58.556 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/mpegts_table.c:107 0x7fd040590864 0x7fd040319000
2017-11-21 00:45:58.590 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/dvb_psi_lib.c:128 (discriminator 3) 0x7fd040581d35 0x7fd040319000
2017-11-21 00:45:58.624 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/dvb_psi_lib.c:173 0x7fd040581f7a 0x7fd040319000
2017-11-21 00:45:58.658 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/mpegts_input.c:1263 (discriminator 2) 0x7fd04057c067 0x7fd040319000
2017-11-21 00:45:58.692 [  ALERT] CRASH: /tvheadend/tvheadend/src/input/mpegts/mpegts_input.c:1623 0x7fd04057d451 0x7fd040319000
2017-11-21 00:45:58.725 [  ALERT] CRASH: /tvheadend/tvheadend/src/wrappers.c:161 0x7fd04047a691 0x7fd040319000
./start_tvh_full_trace: line 2:  3254 Segmentation fault      ./build.linux/tvheadend -l tvh.log --trace all,-tbl-eit,-tbl-base,-scanfile,-epg,-idnode,-idnode,-epggrab,-upnp,-access,-http,-descrambler-emm -c ./test-conf

```
Can't try to fix that now (too late). I guess the fix should be (dvbcam.c/555):
`    if (as->cat_data && memcmp(data, as->cat_data, len) == 0) continue;`
Because the 'as->cat_data' is NULL at first, I guess.

jasmin-j commented:
So the code should be:
```
dvbcam.c:
  TAILQ_FOREACH(as, &dvbcam_active_services, global_link) {
    if (!as->is_ddci) continue;
    if (as->caid == 0) continue;
    if (((mpegts_service_t *)as->td_service)->s_dvb_mux != mux) continue;
    if (as->cat_data && (len == as->cat_data_len) && memcmp(data, as->cat_data, len) == 0) continue;
```

perexg commented:
Fixes:
https://github.com/tvheadend/tvheadend/commit/303683b75cb0b566f9b26704517ec7e2caea7400 , https://github.com/tvheadend/tvheadend/commit/bcad71ba8b9b366686ebf8eb6a43d06eb037d490

jasmin-j commented:
> 303683b

There is 'es' still destroyed (freed) when it gets accessed in the 'for' loop:
```
    for (c = LIST_FIRST(&es->es_caids); c != NULL; c = cn) {
      cn = LIST_NEXT(c, link);
      if (c->delete_me) {
        LIST_REMOVE(c, link);
        free(c);
      }
      if (es->es_delete_me)
        service_stream_destroy(s, es);
    }

```
The for loops over 'es->es_caids', but 'service_stream_destroy(s, es);' will remove the whole list, so 'cn' points to an already deleted object in the next iteration.
I would suggest to change the code to:
```
    if (es->es_delete_me) {
      service_stream_destroy(s, es);
      continue;
    }
    for (c = LIST_FIRST(&es->es_caids); c != NULL; c = cn) {
      cn = LIST_NEXT(c, link);
      if (c->delete_me) {
        LIST_REMOVE(c, link);
        free(c);
      }
    }

```

perexg commented:
@jasmin-j : I don't see any real benefit for continue. I just moved destroy call outside the es_caids loop. https://github.com/tvheadend/tvheadend/commit/fc40b2fe5531b6e91e2c20385b2f9c98149fb22b

herrnst commented:
Todays (current) HEAD changed behaviour: Descrambling still doesn't work, but instead of waiting forever with nothing visible happening, Kodi now quickly complains with "encrypted channel" (regardless of which one I try), TVH logs this:

    2017-11-21 19:02:36.893 [   INFO]:mpegts: 354MHz in Unitymedia NRW - tuning on ST STV0367 DDB DVB-C/T : DVB-C #0
    2017-11-21 19:02:36.893 [  TRACE]:dvbcam: start service 0x556df8b5ba80
    2017-11-21 19:02:36.893 [   INFO]:subscription: 000F: "192.168.0.120 [  | Kodi Media Center ]" subscribing on channel "Fox HD", weight: 150, adapter: "ST STV0367 DDB DVB-C/T : DVB-C #0", network: "Unitymedia NRW", mux: "354MHz", provider: "Unitymedia", service: "Fox HD", profile="htsp", hostname="192.168.0.120", username="192.168.0.120", client="Kodi Media Center"
    2017-11-21 19:02:37.287 [  TRACE]:dvbcam: cannot find active service entry
    2017-11-21 19:02:37.566 [  TRACE]:descrambler: CAT data (len 42)
    2017-11-21 19:02:37.566 [  TRACE]:descrambler: [hexdump]
    2017-11-21 19:02:38.893 [WARNING]:subscription: 000F: service instance is bad, reason: No descrambler
    2017-11-21 19:02:38.893 [  TRACE]:descrambler: mux 0x556df8b00500 - flush tables
    2017-11-21 19:02:40.893 [ NOTICE]:subscription: 000F: No input source available for subscription "192.168.0.120 [  | Kodi Media Center ]" to channel "Fox HD"

Note the "no descrambler" warning (yes, adapter7/ca0 is enabled in the DVB inputs tab and reports "AlphaCrypt" and module ready, aswell as "Config - CAs - Linux DVB CAM").

perexg commented:
Could you try updated master? I accidently reverted one condition.

herrnst commented:
That does CAM assignment again, but shortly after switching to a scrambled service, TVH crashes. gdb bt full in a few moments.

herrnst commented:
GDB: https://pastebin.com/RNvVs1ws
TVH crash dump plus preceeding log: https://pastebin.com/46XYGTPS

perexg commented:
Fixed. Could you retest?

herrnst commented:
Looks like TVH subscribes the first (and only the first) CA PID that the CAM also announces. With the AC, that is

    2017-11-21 20:15:30.375 [   INFO]:en50221: CAM slot 0 supported CAIDs:  1838 (NagraVision) 0648 (Irdeto) 1702 (BetaCrypt) 1722 (BetaCrypt)
    2017-11-21 20:15:30.375 [   INFO]:en50221: CAM slot 0 supported CAIDs:  1762 (BetaCrypt) 4A20 (0x4a20) 0500 (Viaccess) 0B00 (Conax)
    2017-11-21 20:15:30.375 [   INFO]:en50221: CAM slot 0 supported CAIDs:  0100 (Seca) 1833 (NagraVision) 1834 (NagraVision) 0D05 (CryptoWorks)
    2017-11-21 20:15:30.375 [   INFO]:en50221: CAM slot 0 supported CAIDs:  0D95 (CryptoWorks ICE) 09C4 (NDS) 093E (NDS) 0963 (NDS)
    2017-11-21 20:15:30.375 [   INFO]:en50221: CAM slot 0 supported CAIDs:  0919 (NDS) 09C7 (NDS)

If ie. 1722 comes first in the CAT/PAT (you know...) like this (WebUI paste):

    Index | PID | Type | Language | Details
    -- | -- | -- | -- | --
    1 | 0x021f /   543 | MPEG2VIDEO |   |  
    2 | 0x0220 /   544 | MPEG2AUDIO | ger |  
    3 | 0x0226 /   550 | CA |   | CAIDS: 1831:000000
    4 | 0x0227 /   551 | CA |   | CAIDS: 1722:000000
    5 | 0x0228 /   552 | CA |   | CAIDS: 1850:000000
    6 | 0x09f6 /  2550 | CA |   | CAIDS: 1838:000000
    7 | 0x09f8 /  2552 | CA |   | CAIDS: 1835:000000
    8 | 0x1e5d /  7773 | CA |   | CAIDS: 098e:000000

then that one will be subscribed and piped to the CAM. My CAM and SC won't care since it's a 1838 Nagra card. If that map looks like this:

    Index	PID	Type	Language	Details
    1	0x0215 /   533	H264	 	 
    2	0x0216 /   534	MPEG2AUDIO	ger	 
    3	0x021e /   542	CA	 	CAIDS: 1850:000000
    4	0x09ec /  2540	CA	 	CAIDS: 1838:000000
    5	0x1e5b /  7771	CA	 	CAIDS: 098e:000000

then 1838 gets through (first match) and CAM and card react and either return CSA keys and thus a descrambled stream, or it's a notfound plus MMI.

Guess if CAID-PID/CAM evaluation won't stop on the first match but rather subscribe everything, all is good.

jasmin-j commented:
> Looks like TVH subscribes the first (and only the first) CA PID that the CAM also announces.

Yes, this is how it's done. I never changed the lookup code from the original.
For hard wired CAMs this is OK, because the CAM gets all PIDs anyway. So the CAM selection of a hard wired CAM works good.

For a DD CI CAM, we need to subscribe the required CA descriptor PID from the DVB card to feed it via the ciX/secX device into the CAM.

> Guess if CAID-PID/CAM evaluation won't stop on the first match but rather subscribe everything, all is good.

I don't want to do that, because for MTD we need to rewrite all the PIDs from all stream parts including the CAT. So I would like to use the CAID for the CAM selection which is really used. And this CAID is stored in the dvbcam_active_service_t structure and **only** this **single** CAID is currently used to select the right ECM PID. We would need to change the code to select the whole list of ECM PIDs and this is really not required to get the CAM decrypting.

But I don't know if there is anywhere the info which CAID the active service is currently using for decryption.
I hope @perexg can answer this.

perexg commented:
@jasmin-j: The CAID is selected in the first lookup 'check all elementary streams for CAIDs and find CAM'. Indeed, there is an issue for CAMs with multiple CAIDs. Do we have a way to determine which CAID is really used by the CAM (feedback)?

I added code for s_dvb_forcecaid to https://github.com/tvheadend/tvheadend/commit/d85137af3be4f5903c06c8e940a523081cf94826 . So you can manually select the CAID per service (it's not ideal, I know).

We can also add another option with a preferred CAID list to the dvbcam config. Another way (already implemented) is to sort CAID in the elementary stream filters (Configuration / Stream / CA Stream Filters).

The subscription of all PIDs is really bad idea. The hardware (tuner/bridge) has usually limited count of filters (around 32) so if you use 10-16 PIDs for ECM/EMM (and only 2 of them are really used per service), then you have only rest for PAT/PMT/EIT/A/V streams which is really a small number.

herrnst commented:
@perexg @jasmin-j If the CAM just announces one CAID (I don't know what "official" provider modules do, guess they only send the CAID they're intended for), this will work ofc. In the AC/One4All case, users at least have to set the CAM from "MULTI" to "SINGLE" CAID mode (that will result in only having the CAID reported that the inserted card handles), but think about the multicard CAMs (ie. Unicam Dual etc, where people put both their V14 and a HDplus card in), those definitely won't fully work. And if for some reason a provider CAM pops up that also supports (and needs) multiple CAs, this also won't work.

Configuring stream filters isn't a really good option since that's prone to errors and with that will risk that services with other CAIDs won't descramble anymore, and manually configuring stream filters for each service isn't an option at all.

@jasmin-j How does VDR handle this? I can't remember I had to set up anything or even care about PIDs or CAIDs when initially testing and fixing the DD CI hardware during the mainline merge thing, and I didn't have to play with the CAID capability setting of the CAM - that thing just worked with MTD. As VDR does it internally ("outside" of DDCI), I believe Klaus handles all PIDs or at least those that the CAM announces to support.

jasmin-j commented:
> Do we have a way to determine which CAID is really used by the CAM (feedback)?

As @herrnst just said, the AlphaCrypt CAM has a configuration parameter to announce the CAID used by the smart card or all the encryption systems it can decode. The default is the latter, if I remember it correctly.
My old CAM announces only one CAID.

> So you can manually select the CAID per service (it's not ideal, I know).

With that you answered my question, if we can determine the really used CAID from the stream, the answer is NO. But I think the solution not bad. I don't think that the encryption method for a service does change very often.

> We can also add another option with a preferred CAID list to the dvbcam config.

This would solve the issue completely I guess. But as it is with all user configurations this will increase the support questions.

> Another way (already implemented) is to sort CAID in the elementary stream filters

Will not work for CAMs who can really decode more encryption systems.

> @jasmin-j How does VDR handle this?
> I believe Klaus handles all PIDs or at least those that the CAM announces to support.

You answered it already. VDR selects the PIDs for **all** CAIDs a CAM does announce. That's the reason you should configure the AC CAM to announce only the CAIDs the smard card can provide.

> The subscription of all PIDs is really bad idea. The hardware (tuner/bridge) has usually limited count of filters (around 32)

With that said I would suggest to implement the option with a preferred CAID list to the dvbcam config and dvbcam using **all** CAIDs it gets from the CAM. Then it depends on the CAM configuration in the first place and a user may override it in the preferred configuration. And of course the just implemented s_dvb_forcecaid will then be the last filter to select only a single CAID.
@perexg, @herrnst is this a solution which would be useful in your opinion?

herrnst commented:
> [...] the AlphaCrypt CAM has a configuration parameter to announce the CAID used by the smart card or all the encryption systems it can decode. The default is the latter, if I remember it correctly.

Correct. Unfortunately, when only using TVH without any external tools like gnutv, one doesn't have a chance to change the CAM configuration since (as it seems) one can't access the CAM menu, but this is probably subject for a separate feature request, together with matching HTSP and even Kodi integration :-)

> The hardware (tuner/bridge) has usually limited count of filters (around 32)

Errrrr... Isn't that done in the DVB Core? At least in ddbridge I haven't seen anything that would do PID filtering in hardware, can't really say how other devices handle this...

> [...] is this a solution which would be useful in your opinion?

Sounds like an automagic thing, with the user being able to correct things if something goes wrong, so yes. Another thing: I noticed the AC in MULTI mode announces the card CAID as the first one in the supported list, so maybe change the way things are evaluated to go through the CAMs CAID list and match the PIDs accordingly, that'd probably also help.

perexg commented:
Ok, I pushed more fine CAID filtering / selection to https://github.com/tvheadend/tvheadend/commit/9ea470ae6c7928e5c4a368b96bede2c370b9887c . You can handle all CAIDs per CAM now, too. I believe that for the first phase, the major issues are resolved, only MTD implementation is missing.

herrnst commented:
I'm not sure if I get this correctly now. Isn't the "All CAIDs" option supposed to send _ALL_ CAID PIDs to the CAM? If so, something's still missing, because the CAM doesn't decrypt or MMI-complain yet (same with the other two options). Though, if I enter 1838 in the CAID filter list, things just start to work without further ado.

Otherwise, can't see any issue like other scrambled channels stealing the currently active one's session (channel stays active - though hadn't seen that issue before).

jasmin-j commented:
> I believe that for the first phase, the major issues are resolved, only MTD implementation is missing.

Well, phase 2 would be MCD. For that I need to allow a list of services to be attached to DD CI and maybe some changes in dvbcam also
Phase 3 is then MTD.

But first we need to finish the current implementation.
During my review I found an issue with the ALL mode (or I haven't understood how it is intended).
The current ALL code subscribes **all** CAIDs of the **service**, which are also supported by the CAM. In my opinion we had this in a step before, where all ECMs for all CAIDs have been subscribed, even those which are not required by the service.
I would have thought ALL means **all** CAIDs from the **configured list**, which are supported by the CAM.
To get this behaviour we simply need to add:
```
      TAILQ_FOREACH(st, &t->s_filt_components, es_link) {
        if (st->es_type != SCT_CA) continue;
        LIST_FOREACH(c, &st->es_caids, link) {
          if (c->caid != dc->caid_list[i]) continue;     <----- Add this
          if (i >= ARRAY_SIZE(as->caid_list)) {
            tvherror(LS_DVBCAM, "CAID service list overflow");
            break;
          }
```
It is late already and I might be wrong, but when we add the line above I think we can remove the 'if' and the error printing, because both lists have the same size.
And you should add a define for the list size with value 32 and use it in the definition of the lists.

And there is:
```
    if (dc->caid_select == DVBCAM_SEL_LAST && j > 0) {
      as->caid_list[0] = as->caid_list[j-1];
      as->caid_list[1] = 0;   <----- Termination of the list is missing
    }
```
But maybe I haven't understood how LAST should work.

What did you intend with the LAST mode?
I don't understand how I can use this.
What is so special with the last entry of the configured list?
The same behaviour could be achieved by only one entry in the configured list.

perexg commented:
New code: https://github.com/tvheadend/tvheadend/commit/7f81d726af1f92b268d031c08cf5d00f0f77a4c6 ; it contains @jasmin-j review fixes .

The LAST mode is just because I wanted to have more than 2 items in the CAID selection method list :) Don't try to find something special behind it :)

@herrnst : If the CAID filter is empty, all CAIDs should be selected (with 'All CAIDs' settings). I added more log code to dvbcam. Could you show me log for your 'All CAIDs' settings? (--trace dvbcam)

herrnst commented:
@perexg From what I've seen yesterday, this doesn't seem to be the case (and a bit of semi-random debug output sort-of confirmed this), so ATM: CAIDs in the filter list results in PIDs being subscribed which share common CAIDs with what the CAM announced (mode: all), if the filter list is empty, old behaviour applies (ie. first match, first serve). Will post the log in the evening.

herrnst commented:
So, this is the log output with this particular service (WebUI paste "After filtering and reordering") channel with current HEAD (d7cab04a8):

    Index | PID | Type | Language | Details
    1 | 0x021f /   543 | MPEG2VIDEO |   |  
    2 | 0x0220 /   544 | MPEG2AUDIO | ger |  
    3 | 0x0226 /   550 | CA |   | CAIDS: 1831:000000
    4 | 0x0227 /   551 | CA |   | CAIDS: 1722:000000
    5 | 0x0228 /   552 | CA |   | CAIDS: 1850:000000
    6 | 0x09f6 /  2550 | CA |   | CAIDS: 1838:000000
    7 | 0x09f8 /  2552 | CA |   | CAIDS: 1835:000000
    8 | 0x1e5d /  7773 | CA |   | CAIDS: 098e:000000

"All CAIDs" and an empty filter list (1838 announced by the CAM but missing in the PID selection):

    2017-11-23 22:49:07.795 [   INFO]:mpegts: 298MHz in Unitymedia NRW - tuning on ST STV0367 DDB DVB-C/T : DVB-C #0
    2017-11-23 22:49:07.795 [  TRACE]:dvbcam: start service 0x557a247d1860
    2017-11-23 22:49:07.795 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: match CAID 1722 PID 551 (0227)
    2017-11-23 22:49:07.795 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: first CAID 1722 selected
    2017-11-23 22:49:07.795 [ NOTICE]:ddci: CAM ci0 assigned to 0x557a247d1860
    2017-11-23 22:49:07.795 [  TRACE]:descrambler: dvbcam-0-0-1722: key state changed from INIT to READY for "Unitymedia NRW/298MHz/BonGusto"
    2017-11-23 22:49:07.795 [  TRACE]:descrambler: service "Unitymedia NRW/298MHz/BonGusto": 0 descramblers (0 ok 0 failed 0 fatal)
    2017-11-23 22:49:07.795 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: add ECM PID 551 (0227) for CAID 1722
    2017-11-23 22:49:07.795 [   INFO]:subscription: 00FD: "192.168.0.120 [  | Kodi Media Center ]" subscribing on channel "BonGusto", weight: 150, adapter: "ST STV0367 DDB DVB-C/T : DVB-C #0", network: "Unitymedia NRW", mux: "298MHz", provider: "Unitymedia", service: "BonGusto", profile="htsp", hostname="192.168.0.120", username="192.168.0.120", client="Kodi Media Center"
    2017-11-23 22:49:08.191 [  TRACE]:descrambler: CAT data (len 42)
    2017-11-23 22:49:08.191 [  TRACE]:descrambler: [...]
    2017-11-23 22:49:08.191 [  TRACE]:dvbcam: 0x557a247d1860: add EMM PID 5376 (1500) for CAID 1722
    2017-11-23 22:49:08.352 [  TRACE]:en50221: dvbca0: ok_descrambling CAPMT enqueued (only)
    2017-11-23 22:49:08.389 [  TRACE]:en50221: dvbca0: ok_descrambling CAPMT sent (only)
    2017-11-23 22:49:08.389 [  TRACE]:en50221: [...]
    2017-11-23 22:49:09.507 [  TRACE]:ddci: CAM ci0 send buff rd(-> CAM):4355, wr:4355
    2017-11-23 22:49:09.507 [  TRACE]:ddci: CAM ci0 recv rd(CAM ->):23520, wr:23520
    2017-11-23 22:49:09.507 [  TRACE]:ddci: CAM ci0 got 19557 scrambled packets from CAM
    2017-11-23 22:49:13.795 [WARNING]:subscription: 00FD: service instance is bad, reason: No input detected
    2017-11-23 22:49:13.795 [  TRACE]:descrambler: mux 0x557a247b8010 - flush tables
    2017-11-23 22:49:13.795 [  TRACE]:en50221: dvbca0: not_selected CAPMT enqueued (update)
    2017-11-23 22:49:13.795 [ NOTICE]:ddci: CAM ci0 unassigned from 0x557a247d1860
    2017-11-23 22:49:13.796 [  TRACE]:ddci: CAM ci0 send buff rd(-> CAM):17696, wr:17696
    2017-11-23 22:49:13.845 [  TRACE]:en50221: dvbca0: not_selected CAPMT sent (update)
    2017-11-23 22:49:13.845 [  TRACE]:en50221: [...]
    2017-11-23 22:49:15.795 [ NOTICE]:subscription: 00FD: No input source available for subscription "192.168.0.120 [  | Kodi Media Center ]" to channel "BonGusto"

Again, "All CAIDs", with "1722,1831,1838,1850" in the filter list (the MMI at the end is ok since that channel/service isn't subscribed on the card, but it gets it's ECMs):

    2017-11-23 22:56:24.162 [   INFO]:mpegts: 298MHz in Unitymedia NRW - tuning on ST STV0367 DDB DVB-C/T : DVB-C #0
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: start service 0x557a247d1860
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: match CAID 1722 PID 551 (0227)
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: add CAID 1722 to selection
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: add CAID 1838 to selection
    2017-11-23 22:56:24.162 [ NOTICE]:ddci: CAM ci0 assigned to 0x557a247d1860
    2017-11-23 22:56:24.162 [  TRACE]:descrambler: dvbcam-0-0-1722: key state changed from INIT to READY for "Unitymedia NRW/298MHz/BonGusto"
    2017-11-23 22:56:24.162 [  TRACE]:descrambler: service "Unitymedia NRW/298MHz/BonGusto": 0 descramblers (0 ok 0 failed 0 fatal)
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: add ECM PID 551 (0227) for CAID 1722
    2017-11-23 22:56:24.162 [  TRACE]:dvbcam: dvbca0/0x557a247d1860: add ECM PID 2550 (09F6) for CAID 1838
    2017-11-23 22:56:24.162 [   INFO]:subscription: 00FF: "192.168.0.120 [  | Kodi Media Center ]" subscribing on channel "BonGusto", weight: 150, adapter: "ST STV0367 DDB DVB-C/T : DVB-C #0", network: "Unitymedia NRW", mux: "298MHz", provider: "Unitymedia", service: "BonGusto", profile="htsp", hostname="192.168.0.120", username="192.168.0.120", client="Kodi Media Center"
    2017-11-23 22:56:24.714 [  TRACE]:descrambler: CAT data (len 42)
    2017-11-23 22:56:24.714 [  TRACE]:descrambler: [...]
    2017-11-23 22:56:24.714 [  TRACE]:dvbcam: 0x557a247d1860: add EMM PID 5376 (1500) for CAID 1722
    2017-11-23 22:56:24.714 [  TRACE]:dvbcam: 0x557a247d1860: add EMM PID 5379 (1503) for CAID 1838
    2017-11-23 22:56:24.853 [  TRACE]:en50221: dvbca0: ok_descrambling CAPMT enqueued (only)
    2017-11-23 22:56:24.853 [  TRACE]:en50221: dvbca0: ok_descrambling CAPMT sent (only)
    2017-11-23 22:56:24.853 [  TRACE]:en50221: [...]
    2017-11-23 22:56:29.509 [  TRACE]:ddci: CAM ci0 send buff rd(-> CAM):21885, wr:21885
    2017-11-23 22:56:29.509 [  TRACE]:ddci: CAM ci0 recv rd(CAM ->):68544, wr:68544
    2017-11-23 22:56:29.509 [  TRACE]:ddci: CAM ci0 got 55712 scrambled packets from CAM
    2017-11-23 22:56:30.162 [WARNING]:subscription: 00FF: service instance is bad, reason: No input detected
    2017-11-23 22:56:30.162 [  TRACE]:descrambler: mux 0x557a247b8010 - flush tables
    2017-11-23 22:56:30.162 [  TRACE]:en50221: dvbca0: not_selected CAPMT enqueued (update)
    2017-11-23 22:56:30.162 [ NOTICE]:ddci: CAM ci0 unassigned from 0x557a247d1860
    2017-11-23 22:56:30.162 [  TRACE]:ddci: CAM ci0 send buff rd(-> CAM):24723, wr:24723
    2017-11-23 22:56:30.212 [  TRACE]:en50221: dvbca0: not_selected CAPMT sent (update)
    2017-11-23 22:56:30.212 [  TRACE]:en50221: [...]
    2017-11-23 22:56:30.545 [  TRACE]:en50221: dvbca0: 0x00400041 session 6 connecting
    2017-11-23 22:56:30.545 [  TRACE]:en50221: dvbca0: 0x00400041 session 6 connected
    2017-11-23 22:56:30.863 [  TRACE]:en50221: dvbca0: mmi display ctl cb received for slot 0 session_num 6 cmd_id 0x01 mmi_mode 1
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0: MMI list from CAM in the slot 0:
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0:  title:    ▒AlphaCrypt
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0:  subtitle:
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0:  item 1:   Sie haben keine Berechtigung
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0:  item 2:   dieses Programm zu empfangen !
    2017-11-23 22:56:31.204 [ NOTICE]:en50221: dvbca0:  bottom:   Dr▒cken Sie OK
    2017-11-23 22:56:31.521 [  TRACE]:en50221: dvbca0: 0x00400041 session 6 close
    2017-11-23 22:56:32.162 [ NOTICE]:subscription: 00FF: No input source available for subscription "192.168.0.120 [  | Kodi Media Center ]" to channel "BonGusto"

jasmin-j commented:
I fixed that (and more) already!
Will provide a pull request soon.

perexg commented:
@jasmin-j : Sorry, I pushed my change few seconds ago :(

jasmin-j commented:
I will check your changes. I am sure you didn't fix the WEB UI empty list save problem ;)

herrnst commented:
@perexg Quickly rebuilt HEAD, this works as expected now (ie. empty filter list + All CAIDs, CAM gets all PIDs/CAIDs now which match with what's in the service, some CAIDs in the filter list only adds the specified ones, as long as the CAM supports them).

@jasmin-j No, he didn't :)

Great job by both of you of course!

jasmin-j commented:
> Will post the WEB UI save list problem soon.

Done, see #1038.

With that we can close this pull request and finish phase 1 of the DD CI support implementation.

herrnst commented:
@perexg @jasmin-j Even if this PR is closed (and considered done) now, I discovered one related thing while playing a bit with current HEAD as of now:

Had one scrambled channel play for a few minutes, and then switched to another scrambled one (for which no permission/subscription exists though), and TVH crashed with SIGSEGV. That one wasn't easy to reproduce with a -g -Og build and gdb attached, but "luckily", it crashed again. The key seems to be to have a channel play for a little bit, anyway, GDB bt full, thread apply all bt and the resulting tvh.log at:

https://pastebin.com/rA1Cu6W4

jasmin-j commented:
Hi!
Should be fixed by #1040.
LG JJ
